src/hammer-vlsi/defaults.yml

```diff
@@ -277,6 +277,22 @@ vlsi.inputs:
   # SRAM Parameters
   sram_parameters: []
 
+  # Bumps settings struct, Optional
+  # Currently only support rectangular, evenly-spaced grids
+  # Bumps are numbered starting with 1,1 in the lower left
+  # Unassigned bumps will be removed, special name "+NC", leaves bump unconnected
+  # setting top-level bummps to null corresponds to a no bump chip
+  bumps: null
+  # BumpsDefinition struct members
+  # - x (int) - number of bumps in the x dimension
+  # - y (int) - number of bumps in the y dimension
+  # - pitch (float) - pitch of bumps in microns
+  # - cell (str) - Name of bump cell
+  # assignments - List of BumpAssignment structs
+  #  - name (str) - The name of the net being assigned to a bump
+  #  - x (int) - The X coordinate of the bump assigned to this net
+  #  - y (int) - The Y coordinate of the bump assigned to this net
```

Comment:
To avoid using `null` for a subtree of settings since it may be confusing/error-inducing, could we do something more like this instead?
```
  bumps:
    enabled: true # <-- may be inferrable if x*y = 0
    x: 0
    y: 0
    pitch: 0
    cell: ""
    assignments: []
```

Reply:
I think I like `null` better than having an additional `enabled` field. Do you have an example of an error it might cause?

Reply:
One case that comes to mind is that I'm not sure how hammer_config behaves when the parent yml has `null`s in its tree and you want to override subparts of it. In general, while currently hammer setting trees get serialized to maps/dicts, I would be hesistant to reply on it. It's also inconsistent with what we have in the rest of Hammer. It also interacts strangely with static checking of the IR in the future which will help prevent errors.

Reply:
Here's a more concrete example that yields a `KeyError`:

```python
hammer_config.load_config_from_string("""foobar: {'b': 1}""", is_yaml=True)['foobar']
```

`foobar` is an example of this kind of optional tree and if it is present as a dict, trying to grab it directly via `['foobar']` will cause an error as `hammer_config` will expand the dict.

Reply:
I don't quite understand the error that occurs. I used `get_setting` in both the null and populated case and had no issues.
I do think this is a slightly different case than any we have. IMO the solution most similar to other APIs would be to have a `bumps_mode: [auto,none]` key along with the current key, which is still `null` by default but would be gated by the mode key.

Reply:
The `bumps_mode` solution seems fine to me and consistent with the way we've done everything else, and solves @edwardcwang concerns (I think).

Reply:
`bumps_mode` sounds good to me.

To elaborate on the error - I tried with Colin's specific example below and was able to reproduce this type of error:

```python
import hammer_config
mystr = """
"vlsi.inputs.bumps": {
    "x": 14,
    "y": 14,
    "pitch": 200,
    "cell": "MY_REDACTED_BUMP_CELL",
    "assignments": [
        {"name": "reset", "x": 5, "y": 3},
        {"name": "+NC", "x": 5, "y": 4},
        {"name": "VDD", "x": 2, "y": 1},
        {"name": "VSS", "x": 2, "y": 2},
        {"name": "VDD", "x": 2, "y": 3},
        {"name": "VSS", "x": 2, "y": 4},
        {"name": "VDD", "x": 13, "y": 1},
        {"name": "VSS", "x": 13, "y": 2},
        {"name": "VDD", "x": 13, "y": 3},
        {"name": "VSS", "x": 13, "y": 4},
        {"name": "VDD", "x": 2, "y": 14},
        {"name": "VSS", "x": 2, "y": 13},
        {"name": "VDD", "x": 2, "y": 12},
        {"name": "VSS", "x": 2, "y": 11},
        {"name": "VDD", "x": 13, "y": 14},
        {"name": "VSS", "x": 13, "y": 13},
        {"name": "VDD", "x": 13, "y": 12},
        {"name": "VSS", "x": 13, "y": 11}
    ]
}
"""
hammer_config.load_config_from_string(mystr, is_yaml=True)
# {'vlsi.inputs.bumps.x': 14, 'vlsi.inputs.bumps.y': 14, 'vlsi.inputs.bumps.pitch': 200, 'vlsi.inputs.bumps.cell': 'MY_REDACTED_BUMP_CELL', 'vlsi.inputs.bumps.assignments': [{'name': 'reset', 'x': 5, 'y': 3}, {'name': '+NC', 'x': 5, 'y': 4}, {'name': 'VDD', 'x': 2, 'y': 1}, {'name': 'VSS', 'x': 2, 'y': 2}, {'name': 'VDD', 'x': 2, 'y': 3}, {'name': 'VSS', 'x': 2, 'y': 4}, {'name': 'VDD', 'x': 13, 'y': 1}, {'name': 'VSS', 'x': 13, 'y': 2}, {'name': 'VDD', 'x': 13, 'y': 3}, {'name': 'VSS', 'x': 13, 'y': 4}, {'name': 'VDD', 'x': 2, 'y': 14}, {'name': 'VSS', 'x': 2, 'y': 13}, {'name': 'VDD', 'x': 2, 'y': 12}, {'name': 'VSS', 'x': 2, 'y': 11}, {'name': 'VDD', 'x': 13, 'y': 14}, {'name': 'VSS', 'x': 13, 'y': 13}, {'name': 'VDD', 'x': 13, 'y': 12}, {'name': 'VSS', 'x': 13, 'y': 11}], '_config_path': 'unspecified'}
```

As you can see, the dictionary has been parsed by Hammer into a subtree of settings. If I try to grab `vlsi.inputs.bumps` directly I get a KeyError since that key no longer exists:

```
>>> hammer_config.load_config_from_string(mystr, is_yaml=True)["vlsi.inputs.bumps"]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'vlsi.inputs.bumps'
```

Reply:
But `get_setting` handles this case correctly so in what case do we explicitly grab the dictionary?
I don't really care about this but just want to understand when this error shows up.

Reply:
I'll dig into this a bit more. Not sure off the top of my head why `get_setting` is able to grab the dict correctly. e.g. `bumps["assignments"]` to me looks like it's grabbing `bumps` explicitly as a dict

On a related note, are there any (even simple) tests for this PR?

Reply:
Here's a simple test that reproduces the issue:

```python
    def test_bumps(self) -> None:
        """
        Test that HammerTool bump support works.
        """
        import hammer_config

        tech_dir, tech_dir_base = HammerToolTestHelpers.create_tech_dir("dummy28")
        tech_json_filename = os.path.join(tech_dir, "dummy28.tech.json")
        HammerToolTestHelpers.write_tech_json(tech_json_filename)
        tech = self.get_tech(hammer_tech.HammerTechnology.load_from_dir("dummy28", tech_dir))
        tech.cache_dir = tech_dir
        tech.logger = HammerVLSILogging.context("")

        test = DummyTool()
        test.logger = HammerVLSILogging.context("")
        test.run_dir = tempfile.mkdtemp()
        test.technology = tech
        database = hammer_config.HammerDatabase()
        settings = """
{
	"vlsi.inputs.bumps": {
					"x": 14,
					"y": 14,
					"pitch": 200,
					"cell": "MY_REDACTED_BUMP_CELL",
					"assignments": [
						{"name": "reset", "x": 5, "y": 3},
						{"name": "+NC", "x": 5, "y": 4},
						{"name": "VDD", "x": 2, "y": 1},
						{"name": "VSS", "x": 2, "y": 2},
						{"name": "VDD", "x": 2, "y": 3},
						{"name": "VSS", "x": 2, "y": 4},
						{"name": "VDD", "x": 13, "y": 1},
						{"name": "VSS", "x": 13, "y": 2},
						{"name": "VDD", "x": 13, "y": 3},
						{"name": "VSS", "x": 13, "y": 4},
						{"name": "VDD", "x": 2, "y": 14},
						{"name": "VSS", "x": 2, "y": 13},
						{"name": "VDD", "x": 2, "y": 12},
						{"name": "VSS", "x": 2, "y": 11},
						{"name": "VDD", "x": 13, "y": 14},
						{"name": "VSS", "x": 13, "y": 13},
						{"name": "VDD", "x": 13, "y": 12},
						{"name": "VSS", "x": 13, "y": 11}
					]
				}
}
"""
        database.update_project([hammer_config.load_config_from_string(settings, is_yaml=False)])
        test.set_database(database)
        tech.set_database(database)

        print(test.get_bumps())

        # Cleanup
        shutil.rmtree(tech_dir_base)
        shutil.rmtree(test.run_dir)
```

yields

```
======================================================================
ERROR: test_bumps (__main__.HammerToolTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "src/hammer-vlsi/test.py", line 327, in test_bumps
    print(test.get_bumps())
  File "/dev/hammer/src/hammer-vlsi/hammer_vlsi/hammer_tool.py", line 893, in get_bumps
    bumps = self.get_setting("vlsi.inputs.bumps")
  File "/dev/hammer/src/hammer-vlsi/hammer_vlsi/hammer_tool.py", line 725, in get_setting
    return self._database.get_setting(key, nullvalue)
  File "/dev/hammer/src/hammer_config/config_src.py", line 614, in get_setting
    raise KeyError("Key " + key + " is missing")
KeyError: 'Key vlsi.inputs.bumps is missing'

----------------------------------------------------------------------
```

Reply:
Thanks for explaining this! Coping this test into my updates.